### PR TITLE
Add more data to export command

### PR DESCRIPTION
**Pull request by dklimpel**
Lets export:

- user profile
- device list
- connections

Related to: #14883

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

Signed-off-by: Dirk Klimpel [5740567+dklimpel@users.noreply.github.com](mailto:5740567+dklimpel@users.noreply.github.com)

### CODE REVIEW

File: synapse/handlers/admin.py [hunk 1]
```diff
@@ -247,6 +248,21 @@ async def export_user_data(self, user_id: str, writer: "ExfiltrationWriter") ->
                 )
                 writer.write_state(room_id, event_id, state)
 
+        # Get the user profile
+        profile = await self.get_user(UserID.from_string(user_id))
+        assert profile
```
Comment (clokep) [inline synapse/handlers/admin.py:253]: I think previously if you gave a bogus user ID (`@does_not_exist:example.com`) you'd end up with a bunch of empty directories, now you'll get an assertion error? Or do we check somewhere that the user exists?

Regardless I assume this exists to make mypy happy?

```suggestion
        assert profile is not None
```

File: synapse/handlers/admin.py [hunk 2]
```diff
@@ -247,6 +248,21 @@ async def export_user_data(self, user_id: str, writer: "ExfiltrationWriter") ->
...
+
+        # Get all devices the user has
+        devices = await self._device_handler.get_devices_by_user(user_id)
+        writer.write_devices(devices)
+
+        # Get all connections the user has
+        connections = await self.get_whois(UserID.from_string(user_id))
+        writer.write_connections(
+            connections["devices"][""]["sessions"][0]["connections"]
+        )
```
Comment (clokep) [inline synapse/handlers/admin.py:264]: Would calling this "sessions" be clearer?

**Comment by clokep** [comment]
Looks pretty great! 👍 I left a couple of minor comments. It would be nice to document what gets exported somewhere, but doesn't need to be in this PR.

File: synapse/handlers/admin.py (hunk 2 above)
Comment (dklimpel) [reply synapse/handlers/admin.py:264]: Im not sure if session is the right word.
IMO are devices (acces_toke) sessions. But connection are additionally recorded and stored in the database.

**Comment by dklimpel** [comment]
> Looks pretty great! 👍 I left a couple of minor comments. It would be nice to document what gets exported somewhere, but doesn't need to be in this PR.

I have waited for:
- #14883

File: synapse/handlers/admin.py (hunk 2 above)
Comment (clokep) [reply synapse/handlers/admin.py:264]: I'm not really sure of the right word either. 😢 I unfortunately don't have a great solution. The spec around this data seems to imply it should be tied back to devices, but it doesn't seem that Synapse returns it that way?

Comment (dklimpel) [reply synapse/handlers/admin.py:264]: I also have read the [spec](https://spec.matrix.org/v1.5/client-server-api/#get_matrixclientv3adminwhoisuserid) and was irritated. Synapse returns all last connections without devices. The spec refers to devices and their last connections.

The spec has devices (access_token) -> sessions -> connections. It is also unclear to me how a device can have multiple sessions. Connections is clear for me.

File: synapse/handlers/admin.py (hunk 1 above)
Comment (dklimpel) [reply synapse/handlers/admin.py]: Done

**Comment by clokep** [comment]
Putting back in the queue as I seem to have a lot going on at the moment.

File: docs/usage/administration/admin_faq.md [hunk 3]
```diff
@@ -36,8 +42,37 @@ How can I export user data?
...
+to run Synapse:
+
+```console
+poetry run python -m synapse.app.admin_cmd -c <config_file> export-data <user_id> --output-directory <directory_path>
+```
+
+The directory to store the exported data must be empty.
+It can be set with the optional parameter `--output-directory`.
+Defaults to creating a temp directory. This is a subfolder that starts with
+`synapse-exfiltrate` in `/tmp`, `/var/tmp`, or `/usr/tmp`, in that order.
```
Comment (clokep) [inline docs/usage/administration/admin_faq.md:60]: I had some trouble following this:

```suggestion
The directory to store the export data in can be customised with the
`--output-directory` parameter; ensure that the provided directory is
empty. If this parameter is not provided, Synapse defaults to creating
a temporary directory (which starts with "synapse-exfiltrate") in `/tmp`,
`/var/tmp`, or `/usr/tmp`, in that order.
```

**Review by clokep — APPROVED** [review]

**Comment by clokep** [comment]
Thank you @dklimpel, as always! 🎉

**Pull request merged**